File: Sources/HedosCLI/Commands/Rm.swift
import ArgumentParser
import Foundation
import HedosKernel

struct Rm: AsyncParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "rm",
        abstract: "Delete an installed model — files to the Trash, entry off the shelf.")

    @OptionGroup var global: GlobalOptions

    @Argument(help: "Model id or name.")
    var model: String

    @Flag(
        name: [.long, .customShort("y")],
        help: "Actually delete; without it rm only prints what would go.")
    var yes = false

    func run() async throws {
        let kernel = Session.kernel()
        let shelf = try await Session.shelf(kernel)
        let record = try Session.resolve(model, in: shelf)
        let preview: ModelDeletionPreview
        do {
            preview = try await kernel.deletionPreview(record.id)
        } catch {
            throw CLIError(error.localizedDescription)
        }
        guard yes else {
            try dryRun(preview)
            throw CLIError("nothing deleted — re-run with --yes to confirm.")
        }
        let report: ModelDeletionReport
        do {
            report = try await kernel.deleteModel(record.id)
        } catch {
            throw CLIError(error.localizedDescription)
        }
        if global.json {
            try Out.json(
                RmReport(
                    model: report.modelID, name: report.name, kind: report.kind.rawValue,
                    deleted: true, viaDaemon: report.daemonDeleted,
                    paths: report.trashedPaths, bytes: report.freedBytesEstimate))
        } else if report.daemonDeleted {
            Out.line("deleted \(report.name) via the Ollama daemon")
        } else if report.trashedPaths.isEmpty {
            Out.line("forgot \(report.name) — its files were already gone")
        } else {
            let items = report.trashedPaths.count == 1 ? "1 item" : "\(report.trashedPaths.count) items"
            Out.line(
                "moved \(items) to the Trash (~\(ByteFormat.string(report.freedBytesEstimate))) — deleted \(report.name)"
            )
        }
    }

    private func dryRun(_ preview: ModelDeletionPreview) throws {
        if global.json {
            try Out.json(
                RmReport(
                    model: preview.modelID, name: preview.name, kind: preview.kind.rawValue,
                    deleted: false, viaDaemon: preview.viaDaemon,
                    paths: preview.paths, bytes: preview.bytesEstimate))
            return
        }
        Out.line("\(preview.name) · \(preview.kind.rawValue) · ~\(ByteFormat.string(preview.bytesEstimate))")
        if preview.missing {
            Out.line("  files already gone — would only forget the entry")
        } else if preview.viaDaemon {
            Out.line("  asks the Ollama daemon to delete \(preview.name); shared layers stay")
        } else {
            for path in preview.paths {
                Out.line("  \(path)")
            }
        }
    }
}

struct RmReport: Encodable {
    let model: String
    let name: String
    let kind: String
    let deleted: Bool
    let viaDaemon: Bool
    let paths: [String]
    let bytes: Int64
}

File: cli/src/commands/rm.rs
//! `hedos rm <model>` — remove an installed model. Without `-y` it only previews.

use clap::Args;
use runtime::removal::{ModelRemover, OllamaModelRemover, permanent_delete_trasher};

use crate::error::CliError;
use crate::support::output::Out;
use crate::support::session::{self, Session};

/// Arguments for `rm`.
#[derive(Args)]
pub struct RmArgs {
    /// The model to remove (name, alias, or id).
    model: String,
    /// Actually delete the files (without this, only a dry-run preview is shown).
    #[arg(short = 'y', long)]
    yes: bool,
}

/// Run the `rm` command.
pub async fn run(args: RmArgs, out: &Out) -> Result<(), CliError> {
    let session = Session::open()?;
    let shelf = session.shelf(false).await?;
    let record = session::resolve(&args.model, &shelf, None)?;

    if !args.yes {
        let preview = kernel::removal::preview(record);
        out.line(&format!(
            "Would delete {} — {} item(s), ~{} MB. Re-run with -y to confirm.",
            preview.name,
            preview.paths.len(),
            preview.bytes_estimate / 1_000_000,
        ));
        out.json(&serde_json::json!({
            "model": record.id,
            "name": preview.name,
            "paths": preview.paths,
            "bytesEstimate": preview.bytes_estimate,
            "deleted": false,
        }));
        return Ok(());
    }

    let remover = ModelRemover::new(permanent_delete_trasher(), OllamaModelRemover::new());
    let report = remover.remove(record).await?;
    out.line(&format!(
        "Deleted {} — {} item(s), ~{} MB freed{}",
        report.name,
        report.trashed_paths.len(),
        report.freed_bytes_estimate / 1_000_000,
        if report.daemon_deleted {
            " (via the Ollama daemon)"
        } else {
            ""
        },
    ));
    out.json(&serde_json::json!({
        "model": report.model_id,
        "name": report.name,
        "trashedPaths": report.trashed_paths,
        "freedBytesEstimate": report.freed_bytes_estimate,
        "daemonDeleted": report.daemon_deleted,
        "deleted": true,
    }));
    Ok(())
}
